File: src/main/java/com/itszuvalex/femtocraft/managers/research/PlayerResearch.java
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */

package com.itszuvalex.femtocraft.managers.research;

import com.itszuvalex.femtocraft.Femtocraft;
import com.itszuvalex.femtocraft.api.events.EventTechnology;
import com.itszuvalex.femtocraft.api.research.ITechnology;
import com.itszuvalex.femtocraft.network.FemtocraftPacketHandler;
import com.itszuvalex.femtocraft.network.messages.MessageResearchPlayer;
import com.itszuvalex.femtocraft.utils.FemtocraftUtils;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.EnumChatFormatting;
import net.minecraftforge.common.MinecraftForge;

import java.util.Collection;
import java.util.HashMap;
import java.util.logging.Level;

public class PlayerResearch {
    private final static String mapKey = "techMap";
    private final static String techNameKey = "techname";
    private final static String dataKey = "data";

    public final String username;
    private final HashMap<String, ResearchStatus> techStatus;

    public PlayerResearch(String username) {
        this.username = username;
        techStatus = new HashMap<String, ResearchStatus>();
    }

    public Collection<ResearchStatus> getTechnologies() {
        return techStatus.values();
    }

    /**
     * @param name  Name of researchTechnology to mark as researched
     * @param force Pass true if you want the named researchTechnology to be added if it isn't already discovered. This
     *              will bypass discover checks. This will not post an event.
     * @return True if researchTechnology successfully marked as researched. False otherwise.
     */
    public boolean researchTechnology(String name, boolean force) {
        ITechnology rtech = Femtocraft.researchManager()
                .getTechnology(name);
        ResearchStatus tech = techStatus.get(name);
        if (tech == null && !force) {
            return false;
        }

        if (tech == null) {
            techStatus.put(name, new ResearchStatus(name, true));
            discoverNewTechs(rtech, false);
            sync();
            return true;
        }

        EventTechnology.TechnologyResearchedEvent event = new EventTechnology.TechnologyResearchedEvent(
                username, rtech);
        if (!MinecraftForge.EVENT_BUS.post(event)) {
            tech.researched = true;
            EntityPlayerMP player = MinecraftServer.getServer()
                    .getConfigurationManager().func_152612_a(username);
            if (player != null) {
                ITechnology techno = Femtocraft.researchManager()
                        .getTechnology(name);
                if (techno != null) {
                    FemtocraftUtils.sendMessageToPlayer(player, techno.getLevel().getTooltipEnum() + name
                                                                + EnumChatFormatting.RESET
                                                                + " successfully researched.");
                }
            }
            discoverNewTechs(rtech, true);
            sync();

            return true;
        }
        return false;
    }

    // ---------------------------------------------------------

    private void discoverNewTechs(ITechnology discoverer, boolean notify) {
        for (ITechnology t : Femtocraft.researchManager()
                .getTechnologies()) {
            if (t.getPrerequisites() != null) {
                ResearchStatus ts = techStatus.get(t.getName());
                if (ts != null && ts.researched) {
                    continue;
                }

                boolean discovererPrereq = false;
                boolean shouldDiscover = true;
                for (String st : t.getPrerequisites()) {
                    ITechnology pt = Femtocraft.researchManager().getTechnology(st);

                    if (pt == discoverer && discoverer == null) {
                        discovererPrereq = true;
                    }
                    if (pt == null) continue;

                    ResearchStatus rts = techStatus.get(pt.getName());
                    if (rts == null) {
                        shouldDiscover = false;
                        break;
                    }
                    if (!rts.researched) {
                        shouldDiscover = false;
                        break;
                    }
                }

                if (shouldDiscover && discovererPrereq) {
                    discoverTechnology(t.getName());
                    if (notify) {
                        EntityPlayerMP player = MinecraftServer.getServer()
                                .getConfigurationManager()
                                .func_152612_a(username);
                        if (player != null) {
                            ITechnology techno = Femtocraft.researchManager()
                                    .getTechnology(t.getName());
                            if (techno != null) {
                                FemtocraftUtils.sendMessageToPlayer(player, "New technology "
                                                                            + techno.getLevel().getTooltipEnum()
                                                                            + t.getName() + EnumChatFormatting.RESET
                                                                            + " discovered.");
                            }
                        }
                    }
                }
            }
        }
    }

    public boolean discoverTechnology(String name) {
        if (techStatus.containsKey(name)) return true;

        EventTechnology.TechnologyDiscoveredEvent event = new EventTechnology.TechnologyDiscoveredEvent(
                username, Femtocraft.researchManager().getTechnology(name));
        if (!MinecraftForge.EVENT_BUS.post(event)) {
            techStatus.put(name, new ResearchStatus(name));
            sync();
            return true;
        }
        return false;
    }

    public ResearchStatus getTechnology(String name) {
        return techStatus.get(name);
    }

    public ResearchStatus removeTechnology(String name) {
        return techStatus.remove(name);
    }

    public boolean canDiscoverTechnology(ITechnology tech) {
        if (tech.getPrerequisites() == null) {
            return true;
        }

        for (String str : tech.getPrerequisites()) {
            ITechnology prereq = Femtocraft.researchManager().getTechnology(str);
            if (prereq == null) continue;
            ResearchStatus ts = techStatus.get(prereq.getName());
            if (ts == null) {
                return false;
            }
            if (!ts.researched) {
                return false;
            }
        }

        return true;
    }

    public boolean hasDiscoveredTechnology(ITechnology tech) {
        return hasDiscoveredTechnology(tech.getName());
    }

    public boolean hasDiscoveredTechnology(String tech) {
        ResearchStatus ts = techStatus.get(tech);
        return ts != null;
    }

    public boolean hasResearchedTechnology(ITechnology tech) {
        return tech == null || hasResearchedTechnology(tech.getName());
    }

    public boolean hasResearchedTechnology(String tech) {
        if (tech == null || tech.equals("")) {
            return false;
        }

        ResearchStatus ts = techStatus.get(tech);
        return ts != null && ts.researched;
    }

    public void sync() {
        EntityPlayerMP player = MinecraftServer.getServer()
                .getConfigurationManager().func_152612_a(username);
        if (player == null) {
            return;
        }

        sync(player);
    }

    // ---------------------------------------------------------

    public void sync(EntityPlayerMP player) {
        Femtocraft.log(Level.FINE, "Sending research data to player: " + player.getCommandSenderName());
        FemtocraftPacketHandler.INSTANCE().sendTo(new MessageResearchPlayer(this), player);
    }

    public void saveToNBTTagCompound(NBTTagCompound compound) {
        NBTTagList list = new NBTTagList();

        for (ResearchStatus status : techStatus.values()) {
            NBTTagCompound cs = new NBTTagCompound();
            cs.setString(techNameKey, status.tech);

            NBTTagCompound data = new NBTTagCompound();
            status.saveToNBTTagCompound(data);

            cs.setTag(dataKey, data);
            list.appendTag(cs);
        }

        compound.setTag(mapKey, list);
    }

    // ---------------------------------------------------------

    public void loadFromNBTTagCompound(NBTTagCompound compound) {
        NBTTagList list = compound.getTagList(mapKey, 10);

        for (int i = 0; i < list.tagCount(); ++i) {
            NBTTagCompound cs = (NBTTagCompound) list.getCompoundTagAt(i);
            String techname = cs.getString(techNameKey);

            NBTTagCompound data = cs.getCompoundTag(dataKey);
            ResearchStatus status = new ResearchStatus(
                    techname);
            status.loadFromNBTTagCompound(data);

            techStatus.put(techname, status);
        }
    }

    public static class TechnologyNotFoundException extends Exception {
        public String errMsg;

        public TechnologyNotFoundException(String message) {
            errMsg = message;
        }
    }
}

File: src/main/scala/com/itszuvalex/femtocraft/managers/research/PlayerResearch.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.managers.research


import java.io.File
import java.util
import java.util.UUID

import com.itszuvalex.femtocraft.Femtocraft
import com.itszuvalex.femtocraft.api.core.ISaveable
import com.itszuvalex.femtocraft.api.events.EventTechnology
import com.itszuvalex.femtocraft.api.implicits.SugarImplicits._
import com.itszuvalex.femtocraft.api.managers.research.{IPlayerResearch, IResearchStatus}
import com.itszuvalex.femtocraft.api.research.ITechnology
import com.itszuvalex.femtocraft.api.utils.FemtocraftUtils
import com.itszuvalex.femtocraft.configuration.XMLLoaderWriter
import com.itszuvalex.femtocraft.managers.research.PlayerResearch._
import com.itszuvalex.femtocraft.network.FemtocraftPacketHandler
import com.itszuvalex.femtocraft.network.messages.MessageResearchPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.nbt.{NBTTagCompound, NBTTagList}
import net.minecraft.util.EnumChatFormatting
import net.minecraftforge.common.MinecraftForge
import org.apache.logging.log4j.Level

import scala.collection.JavaConversions._
import scala.collection.mutable

object PlayerResearch {
  private val mapKey      = "techMap"
  private val techNameKey = "techname"
  private val dataKey     = "data"
}

class PlayerResearch(private val uuid: UUID) extends IPlayerResearch with ISaveable {
  private val techStatus           = new mutable.HashMap[String, ResearchStatus]
  private var xml: XMLLoaderWriter = null

  def setFile(file: File): Unit = {
    xml = new XMLLoaderWriter(file)
    load()
  }

  /**
   * @param name  Name of researchTechnology to mark as researched
   * @param force Pass true if you want the named researchTechnology to be added if it isn't already discovered. This
   *              will bypass discover checks. This will not post an event.
   * @return True if researchTechnology successfully marked as researched. False otherwise.
   */
  override def researchTechnology(name: String, force: Boolean): Boolean = researchTechnology(name, force, true)

  override def researchTechnology(name: String, force: Boolean, notify: Boolean): Boolean = {
    val rtech = Femtocraft.researchManager.getTechnology(name)
    if (rtech == null) return true
    val tech = techStatus.get(name).orNull
    if (tech == null && !force) {
      return false
    }
    if (tech == null) {
      techStatus.put(name, new ResearchStatus(name, true))
      if (notify) {
        val player = FemtocraftUtils.getServerPlayer(uuid)
        if (player != null) {
          val techno = Femtocraft.researchManager.getTechnology(name)
          if (techno != null) {
            FemtocraftUtils
            .sendMessageToPlayer(player,
                                 techno.getLevel.getTooltipEnum + name + EnumChatFormatting
                                                                         .RESET + " successfully researched.")
          }
        }
      }
      discoverNewTechs(false)
      save()
      sync()
      return true
    }
    val event = new EventTechnology.Researched(uuid, rtech)
    if (!MinecraftForge.EVENT_BUS.post(event)) {
      tech.researched = true
      if (notify) {
        val player = FemtocraftUtils.getServerPlayer(uuid)
        if (player != null) {
          val techno = Femtocraft.researchManager.getTechnology(name)
          if (techno != null) {
            FemtocraftUtils
            .sendMessageToPlayer(player,
                                 techno.getLevel.getTooltipEnum + name + EnumChatFormatting
                                                                         .RESET + " successfully researched.")
          }
        }
      }
      discoverNewTechs(true)
      save()
      sync()
      return true
    }
    false
  }

  def discoverNewTechs(notify: Boolean): Boolean = {
    Femtocraft.researchManager.getTechnologies.filterNot(hasDiscoveredTechnology).filterNot(hasResearchedTechnology)
    .foreach { case t if t.isResearchedByDefault => return researchTechnology(t.getName, true, notify)
    case t if t.isDiscoveredByDefault => discoverTechnology(t.getName, notify)
    case t if canDiscoverTechnology(t) => discoverTechnology(t.getName, notify)
    case _ =>
             }
    true
  }

  override def discoverTechnology(name: String): Boolean = discoverTechnology(name, true)

  override def discoverTechnology(name: String, notify: Boolean): Boolean = {
    if (techStatus.containsKey(name)) return true
    val event = new EventTechnology.Discovered(uuid, Femtocraft.researchManager.getTechnology(name))
    if (!MinecraftForge.EVENT_BUS.post(event)) {
      techStatus.put(name, new ResearchStatus(name))
      if (notify) {
        val player = FemtocraftUtils.getServerPlayer(uuid)
        if (player != null) {
          val techno = Femtocraft.researchManager.getTechnology(name)
          if (techno != null) {
            FemtocraftUtils
            .sendMessageToPlayer(player,
                                 "New technology " + techno.getLevel.getTooltipEnum + name + EnumChatFormatting
                                                                                             .RESET + " discovered.")
          }
        }
      }
      save()
      sync()
      return true
    }
    false
  }

  override def sync() = {
    val player = FemtocraftUtils.getServerPlayer(uuid)
    player IfNotNull sync(player)
  }

  override def sync(player: EntityPlayerMP) {
    if (ManagerResearch.debugMessages) Femtocraft.log(Level.INFO, "Sending research data to player: " + player.getCommandSenderName)
    FemtocraftPacketHandler.INSTANCE.sendTo(new MessageResearchPlayer(this), player)
  }

  override def removeTechnology(name: String): ResearchStatus = techStatus.remove(name).orNull

  override def canDiscoverTechnology(tech: ITechnology): Boolean = {
    if (tech.getPrerequisites == null || tech.getPrerequisites.size() == 0) {
      return true
    }
    tech.getPrerequisites.map(Femtocraft.researchManager.getTechnology).forall(hasResearchedTechnology)
  }

  override def hasDiscoveredTechnology(tech: ITechnology): Boolean = tech == null || hasDiscoveredTechnology(tech
                                                                                                             .getName)

  override def hasDiscoveredTechnology(tech: String): Boolean = techStatus.get(tech).orNull != null

  override def hasResearchedTechnology(tech: ITechnology): Boolean = tech == null || hasResearchedTechnology(tech
                                                                                                             .getName)

  override def hasResearchedTechnology(tech: String): Boolean = {
    if (tech == null || (tech == "")) {
      return true
    }
    val ts = techStatus.get(tech).orNull
    ts != null && ts.researched
  }

  override def saveToNBT(compound: NBTTagCompound) {
    val list = new NBTTagList
    techStatus.values.foreach { status =>
      val cs = new NBTTagCompound
      cs.setString(techNameKey, status.tech)
      val data = new NBTTagCompound
      status.saveToNBT(data)
      cs.setTag(dataKey, data)
      list.appendTag(cs)
                              }
    compound.setTag(mapKey, list)
  }

  override def loadFromNBT(compound: NBTTagCompound): Unit = {
    techStatus.clear()

    val list = compound.getTagList(mapKey, 10)
    (0 until list.tagCount).map(list.getCompoundTagAt).foreach { cs =>
      val techname = cs.getString(techNameKey)
      val data = cs.getCompoundTag(dataKey)
      val status = new ResearchStatus(techname)
      status.loadFromNBT(data)
      techStatus.put(techname, status)
                                                               }
  }

  def save(): Unit = {
    if (xml == null) {
      if (ManagerResearch.debugMessages) Femtocraft.log(Level.INFO, "No xml loader found when attempting to save.")
      return
    }
    xml.xml = <xml>
      {for (techStatus <- techStatus) yield techStatus._2.saveAsNode}
    </xml>
    xml.save()
    if (ManagerResearch.debugMessages) Femtocraft.log(Level.INFO, "Saving " + uuid + "'s research data to " + xml.file.getPath + ".")
  }

  def load(): Unit = {
    if (xml == null) {
      if (ManagerResearch.debugMessages) Femtocraft.log(Level.INFO, "No xml loader found when attempting to load.")
      return
    }
    if (ManagerResearch.debugMessages) Femtocraft.log(Level.INFO, "Loading " + uuid + "'s research data from " + xml.file.getPath + ".")
    techStatus.clear()
    xml.load()
    (xml.xml \ ResearchStatus.XMLTag).foreach(node => {
      val status = ResearchStatus(node)
      techStatus.put(status.getTechName, status)
    })
    discoverNewTechs(false)
    sync()
  }

  /**
   *
   * @return Username of the player associated with this research.
   */
  override def getUsername = Femtocraft.uuidManager.getUsername(uuid)

  /**
   *
   * @return All research statuses mapped for this player.
   */
  override def getTechnologies: util.Collection[IResearchStatus] = techStatus.values

  /**
   *
   * @param name
   * @return Returns the IResearchStatus struct from this object for the given technology.
   */
  override def getTechnology(name: String): ResearchStatus = techStatus.get(name).orNull

  /**
   *
   * @return Return uuid of the player associated with this research.
   */
  override def getUUID: UUID = uuid
}
